File: sdk-typescript/src/facilitator-client.ts
/**
 * Veridex TypeScript SDK - Facilitator Client
 * License: Apache-2.0
 *
 * Client for x402 Facilitator payment settlement.
 */

import { createEd25519Signer, getUsdcAddress } from "@x402/stellar";
import { ExactStellarScheme } from "@x402/stellar/exact/client";
import type { PaymentRequirements } from "@x402/core/types";
import type { X402PaymentRequest, X402PaymentResponse, FacilitatorScheme, SmartAccountSigner } from "./types.js";

/**
 * Facilitator Client Configuration
 */
export interface FacilitatorClientConfig {
  /** Facilitator service URL */
  facilitatorUrl: string;

  /** Stellar network */
  network: "pubnet" | "testnet" | "futurenet";

  /** Client secret key for standard Ed25519 signing */
  clientSecretKey?: string;

  /** Custom Smart Account / Passkey signer (for C... contract IDs) */
  customSigner?: SmartAccountSigner;

  /** Custom authorizeEntry function override */
  authorizeEntry?: (
    entry: any,
    signer: any,
    expiration: number,
    networkPassphrase?: string
  ) => Promise<any>;

  /** Request timeout (ms) */
  timeout?: number;

  /** Require the facilitator to advertise sponsored fees before paying. */
  requireSponsoredFees?: boolean;
}

/**
 * Facilitator Client
 *
 * Execute x402 payments via Stellar facilitator.
 */
export class FacilitatorClient {
  private config: FacilitatorClientConfig;
  private exactScheme?: ExactStellarScheme;

  constructor(config: FacilitatorClientConfig) {
    this.config = {
      ...config,
      clientSecretKey: config.clientSecretKey || "",
      timeout: config.timeout || 30000,
      requireSponsoredFees: config.requireSponsoredFees ?? false,
    };

    if (config.customSigner) {
      // Initialize with Smart Account signer (supports C... addresses and custom authorizeEntry)
      const options: Record<string, unknown> = {};
      if (config.authorizeEntry || config.customSigner.authorizeEntry) {
        options.authorizeEntry = config.authorizeEntry || config.customSigner.authorizeEntry;
      }
      this.exactScheme = new ExactStellarScheme(config.customSigner as any, options as any);
    } else if (config.clientSecretKey) {
      const network = this.getNetwork();
      this.exactScheme = new ExactStellarScheme(createEd25519Signer(config.clientSecretKey, network));
    }
  }

  /**
   * Get supported payment schemes
   *
   * @returns Supported schemes
   */
  async getSupportedSchemes(): Promise<{
    kinds: FacilitatorScheme[];
    extensions: string[];
    signers: Record<string, string[]>;
  }> {
    const url = new URL("/supported", this.config.facilitatorUrl);

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to get supported schemes: ${response.statusText}`);
    }

    return (await response.json()) as {
      kinds: FacilitatorScheme[];
      extensions: string[];
      signers: Record<string, string[]>;
    };
  }

  /**
   * Fetch compute capability descriptor (x402ccd/0) per proposal #3117
   */
  async getCapabilityDescriptor(): Promise<import("./types.js").ComputeCapabilityDescriptor> {
    const url = new URL("/.well-known/x402", this.config.facilitatorUrl);
    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to fetch capability descriptor: ${response.statusText}`);
    }

    return (await response.json()) as import("./types.js").ComputeCapabilityDescriptor;
  }

  /**
   * Pay for resource access
   *
   * Creates, signs, and submits x402 payment transaction.
   *
   * @param request - Payment request
   * @returns Payment response
   */
  async pay(request: X402PaymentRequest): Promise<X402PaymentResponse> {
    if (!this.exactScheme) {
      throw new Error("Client secret key not configured");
    }

    // Get facilitator info
    const supported = await this.getSupportedSchemes();
    const network = this.getNetwork();
    const stellarScheme = supported.kinds.find(
      (kind) => kind.scheme === "exact" && kind.network === network
    );

    if (!stellarScheme) {
      throw new Error(`Exact Stellar scheme is not supported for ${network}`);
    }

    const areFeesSponsored = stellarScheme.extra?.areFeesSponsored === true;
    if (this.config.requireSponsoredFees && !areFeesSponsored) {
      throw new Error("Facilitator does not advertise sponsored fees for this network");
    }

    const paymentRequirements: PaymentRequirements = {
      scheme: "exact",
      network,
      asset: request.asset || getUsdcAddress(network),
      amount: request.amountStroops,
      payTo: request.payTo,
      maxTimeoutSeconds: 60,
      extra: { areFeesSponsored },
    };
    const created = await this.exactScheme.createPaymentPayload(2, paymentRequirements);
    const paymentPayload = {
      ...created,
      resource: { url: request.resourceUrl || "" },
      accepted: paymentRequirements,
    };

    const url = new URL("/settle", this.config.facilitatorUrl);

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

    try {
      const response = await fetch(url.toString(), {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ paymentPayload, paymentRequirements }),
        signal: controller.signal,
      });

      if (!response.ok) {
        const error = (await response.json().catch(() => ({}))) as Record<string, any>;
        return {
          status: "error",
          error: error.errorMessage || error.error || `Payment failed: ${response.statusText}`,
          errorCode: error.errorReason || error.errorCode || "UNKNOWN",
          extra: error.extra,
        };
      }

      const result = (await response.json()) as Record<string, any>;
      return {
        status: result.success ? "success" : "error",
        transactionHash: result.transaction,
        ledger: result.ledger,
        error: result.errorMessage,
        errorCode: result.errorReason,
        receipt: result.receipt,
        extra: result.extra,
      };
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Verify a transaction without submitting
   *
   * @param transactionXdr - Transaction XDR
   * @param requirements - Canonical payment requirements used to build the transaction
   * @returns Verification result
   */
  async verify(transactionXdr: string, requirements: PaymentRequirements): Promise<{
    isValid: boolean;
    invalidReason?: string;
    invalidMessage?: string;
    payer?: string;
  }> {
    const paymentPayload = {
      x402Version: 2,
      resource: { url: "" },
      accepted: requirements,
      payload: { transaction: transactionXdr },
    };

    const url = new URL("/verify", this.config.facilitatorUrl);

    const response = await fetch(url.toString(), {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ paymentPayload, paymentRequirements: requirements }),
    });

    if (!response.ok) {
      const error = (await response.json().catch(() => ({}))) as Record<string, any>;
      return {
        isValid: false,
        invalidReason: error.invalidReason || "verification_failed",
        invalidMessage:
          error.invalidMessage || error.error || `Verification failed: ${response.statusText}`,
      };
    }

    return (await response.json()) as any;
  }

  /**
   * Get transaction status
   *
   * @param transactionHash - Transaction hash
   * @returns Transaction status
   */
  async getTransactionStatus(transactionHash: string): Promise<{
    found: boolean;
    successful?: boolean;
    ledger?: number;
    createdAt?: string;
  }> {
    const url = new URL(`/transaction/${transactionHash}`, this.config.facilitatorUrl);

    const response = await fetch(url.toString());

    if (response.status === 404) {
      return { found: false };
    }

    if (!response.ok) {
      throw new Error(`Transaction lookup failed: ${response.statusText}`);
    }

    return (await response.json()) as any;
  }

  /**
   * Get service health status
   *
   * @returns Health status
   */
  async health(): Promise<{
    status: string;
    timestamp: number;
    channels: any;
    network: string;
  }> {
    const url = new URL("/health", this.config.facilitatorUrl);

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Health check failed: ${response.statusText}`);
    }

    return (await response.json()) as any;
  }

  /**
   * Get service statistics
   *
   * @returns Service stats
   */
  async stats(): Promise<{
    uptime: number;
    verifications: any;
    settlements: any;
    channels: any;
    timestamp: number;
  }> {
    const url = new URL("/stats", this.config.facilitatorUrl);

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Stats request failed: ${response.statusText}`);
    }

    return (await response.json()) as any;
  }

  private getNetwork(): "stellar:pubnet" | "stellar:testnet" {
    if (this.config.network === "futurenet") {
      throw new Error("Canonical @x402/stellar v2 supports testnet and pubnet only");
    }
    return this.config.network === "pubnet" ? "stellar:pubnet" : "stellar:testnet";
  }
}

/**
 * Helper to create a Smart Account / Contract (`C...`) signer for Soroban custom authorization
 *
 * @param address - Smart Account contract address (C...) or public key
 * @param signAuthEntryFn - Function signing Soroban authorization entries returning custom ScVal / signature
 * @param authorizeEntryFn - Optional custom authorizeEntry override
 */
export function createSmartAccountSigner(
  address: string,
  signAuthEntryFn: (entryXdr: string) => Promise<{ signedAuthEntry?: string; signatureScVal?: any }>,
  authorizeEntryFn?: (
    entry: any,
    signer: any,
    expiration: number,
    networkPassphrase?: string
  ) => Promise<any>
): SmartAccountSigner {
  return {
    address,
    signAuthEntry: signAuthEntryFn,
    authorizeEntry: authorizeEntryFn,
  };
}

/**
 * Create Facilitator client
 */
export function createFacilitatorClient(config: FacilitatorClientConfig): FacilitatorClient {
  return new FacilitatorClient(config);
}

File: sdk-python/veridex/facilitator_client.py
"""
Veridex Python SDK - Facilitator Client
License: Apache-2.0

Client for x402 Facilitator payment settlement.
"""

import requests
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
from .types import PaymentResponse, FacilitatorScheme


@dataclass
class PaymentRequest:
    """A canonical x402 payload and the requirements it was signed against."""
    resource_url: str
    payment_payload: Dict[str, Any]
    payment_requirements: Dict[str, Any]
    tool_name: Optional[str] = None
    session_id: Optional[str] = None


class FacilitatorClient:
    """
    Facilitator Client

    Execute x402 payments via Stellar facilitator.

    Args:
        facilitator_url: Facilitator service URL
        network: Stellar network (pubnet, testnet, futurenet)
        client_secret_key: Client Stellar secret key for signing
        timeout: Request timeout in seconds

    Example:
        >>> client = FacilitatorClient(
        ...     facilitator_url="http://localhost:3002",
        ...     network="testnet",
        ...     client_secret_key="S..."
        ... )
        >>> payment = client.pay(PaymentRequest(
        ...     resource_url="https://api.example.com/tool",
        ...     amount_stroops="100000"
        ... ))
        >>> if payment.status == "success":
        ...     print(f"TX: {payment.transaction_hash}")
    """

    def __init__(
        self,
        facilitator_url: str,
        network: str = "testnet",
        client_secret_key: Optional[str] = None,
        timeout: int = 30,
    ):
        self.facilitator_url = facilitator_url.rstrip("/")
        self.network = network
        self.timeout = timeout
        self.session = requests.Session()

        if network not in ("pubnet", "testnet"):
            raise ValueError(f"Invalid network: {network}")

    def get_supported_schemes(self) -> List[FacilitatorScheme]:
        """
        Get supported payment schemes

        Returns:
            List of supported schemes
        """
        url = f"{self.facilitator_url}/supported"
        response = self.session.get(url, timeout=self.timeout)
        response.raise_for_status()

        return response.json().get("kinds", [])

    def pay(self, request: PaymentRequest) -> PaymentResponse:
        """
        Pay for resource access

        Submits a canonical payload created by an x402 Stellar signer.

        Args:
            request: Payment request

        Returns:
            Payment response with transaction hash or error

        Raises:
            ValueError: If client secret key not configured
            requests.RequestException: If request fails
        """
        url = f"{self.facilitator_url}/settle"

        response = self.session.post(
            url,
            json={
                "x402Version": request.payment_payload.get("x402Version", 2),
                "paymentPayload": request.payment_payload,
                "paymentRequirements": request.payment_requirements,
            },
            timeout=self.timeout,
        )

        data = response.json()

        if response.ok:
            return PaymentResponse(
                status=data.get("success") and "success" or "error",
                transaction_hash=data.get("transaction" ) or data.get("transactionHash"),
                ledger=data.get("ledger"),
            )
        else:
            return PaymentResponse(
                status="error",
                error=data.get("errorMessage") or data.get("error") or response.reason,
                error_code=data.get("errorReason") or data.get("errorCode") or "UNKNOWN",
            )

    def verify(
        self, payment_payload: Dict[str, Any], payment_requirements: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Verify a transaction without submitting

        Args:
            transaction_xdr: Transaction XDR
            expected_amount: Expected amount in stroops

        Returns:
            Verification result
        """
        url = f"{self.facilitator_url}/verify"

        response = self.session.post(
            url, json={"paymentPayload": payment_payload, "paymentRequirements": payment_requirements}, timeout=self.timeout
        )

        return response.json()

    def get_transaction_status(self, transaction_hash: str) -> Dict[str, Any]:
        """
        Get transaction status

        Args:
            transaction_hash: Transaction hash

        Returns:
            Transaction status
        """
        url = f"{self.facilitator_url}/transaction/{transaction_hash}"
        response = self.session.get(url, timeout=self.timeout)

        if response.status_code == 404:
            return {"found": False}

        response.raise_for_status()
        return response.json()

    def health(self) -> Dict[str, Any]:
        """
        Get service health status

        Returns:
            Health status dictionary
        """
        url = f"{self.facilitator_url}/health"
        response = self.session.get(url, timeout=self.timeout)
        response.raise_for_status()
        return response.json()

    def stats(self) -> Dict[str, Any]:
        """
        Get service statistics

        Returns:
            Service stats dictionary
        """
        url = f"{self.facilitator_url}/stats"
        response = self.session.get(url, timeout=self.timeout)
        response.raise_for_status()
        return response.json()

    def close(self):
        """Close the HTTP session"""
        self.session.close()

    def __enter__(self):
        """Context manager entry"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.close()
